### PR TITLE
`ismp parachain` benchmark

### PR DESCRIPTION
**Pull request by dharjeezy**
This PR benchmarks the ISMP Parachain Pallet and the inclusion to all supported runtimes.

closes https://github.com/polytope-labs/hyperbridge/issues/349

### CODE REVIEW

File: modules/ismp/clients/parachain/client/src/benchmarking.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+			Pallet::<T>::add_parachain(RawOrigin::Root.into(), state_machines)?;
+			Pallet::<T>::remove_parachain(RawOrigin::Root.into(), vec![0, 1, 2, 3, 4])?;
+		}
+
+		Ok(())
+	}
+
+	#[benchmark]
+	fn update_parachain_consensus() -> Result<(), BenchmarkError> {
+		let consensus_message = ConsensusMessage {
```
Comment (Wizdave97) [inline modules/ismp/clients/parachain/client/src/benchmarking.rs:56]: Should use a valid consensus proof so we can have a rough estimation for the execution weight.

File: modules/ismp/clients/parachain/client/src/benchmarking.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+use frame_benchmarking::v2::*;
+use frame_system::RawOrigin;
+
+#[benchmarks]
+mod benchmarks {
+	use super::*;
+	use ismp::messaging::{ConsensusMessage, Message};
+
+	#[benchmark]
+	fn add_parachain() -> Result<(), BenchmarkError> {
```
Comment (seunlanlege) [inline modules/ismp/clients/parachain/client/src/benchmarking.rs:28]: this should be dynamic

File: modules/ismp/clients/parachain/client/src/weights.rs
Comment (seunlanlege) [inline modules/ismp/clients/parachain/client/src/weights.rs:1]: this file is unneeded

File: parachain/runtimes/gargantua/src/ismp.rs [hunk 3]
```diff
@@ -141,6 +141,7 @@ impl pallet_call_decompressor::Config for Runtime {
 impl ismp_parachain::Config for Runtime {
 	type RuntimeEvent = RuntimeEvent;
 	type IsmpHost = Ismp;
+	type WeightInfo = ismp_parachain::weights::WeightInfo<Runtime>;
```
Comment (seunlanlege) [inline parachain/runtimes/gargantua/src/ismp.rs:144]: weights are runtime dependent

File: modules/ismp/clients/parachain/client/src/benchmarking.rs (hunk 1 above)
Comment (dharjeezy) [reply modules/ismp/clients/parachain/client/src/benchmarking.rs:60]: This is now addressed.

File: modules/ismp/clients/parachain/client/src/benchmarking.rs (hunk 2 above)
Comment (dharjeezy) [reply modules/ismp/clients/parachain/client/src/benchmarking.rs]: this is now addressed

File: modules/ismp/clients/parachain/client/src/weights.rs
Comment (dharjeezy) [reply modules/ismp/clients/parachain/client/src/weights.rs:1]: just included only the weights there

File: modules/ismp/clients/parachain/client/src/weights.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+#![cfg_attr(rustfmt, rustfmt_skip)]
+#![allow(unused_parens)]
+#![allow(unused_imports)]
+#![allow(missing_docs)]
+
+use frame_support::{traits::Get, weights::{Weight, constants::RocksDbWeight}};
+use core::marker::PhantomData;
+
+pub trait WeightInfoTrait {
```
Comment (Wizdave97) [inline modules/ismp/clients/parachain/client/src/weights.rs]:
```suggestion
pub trait WeightInfoT {
```

File: modules/ismp/clients/parachain/client/src/weights.rs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    fn add_parachain() -> Weight;
+    fn remove_parachain() -> Weight;
+    fn update_parachain_consensus() -> Weight;
+
+}
+/// Weight functions for `ismp_parachain`.
```
Comment (Wizdave97) [inline modules/ismp/clients/parachain/client/src/weights.rs]: Only the trait definition is necessary

File: modules/ismp/clients/parachain/client/src/lib.rs [hunk 6]
```diff
@@ -62,6 +67,8 @@ pub mod pallet {
 
 		/// The underlying [`IsmpHost`] implementation
 		type IsmpHost: IsmpHost + Default;
+		/// WeightInfo
+		type WeightInfo: WeightInfoTrait;
 	}
```
Comment (Wizdave97) [inline modules/ismp/clients/parachain/client/src/lib.rs:73]: Apply the weights to the extrinsics

File: modules/ismp/clients/parachain/client/src/weights.rs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+#![cfg_attr(rustfmt, rustfmt_skip)]
+#![allow(unused_parens)]
+#![allow(unused_imports)]
+#![allow(missing_docs)]
+
+use frame_support::{traits::Get, weights::{Weight, constants::RocksDbWeight}};
+use core::marker::PhantomData;
+
+pub trait WeightInfoTrait {
+    fn add_parachain() -> Weight;
```
Comment (Wizdave97) [inline modules/ismp/clients/parachain/client/src/weights.rs]:
```suggestion
    fn add_parachain(n: u32) -> Weight;
```

File: modules/ismp/clients/parachain/client/src/weights.rs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    fn remove_parachain() -> Weight;
```
Comment (Wizdave97) [inline modules/ismp/clients/parachain/client/src/weights.rs]:
```suggestion
    fn remove_parachain(n: u32) -> Weight;
```

**Review by Wizdave97 — CHANGES_REQUESTED** [review]

File: parachain/runtimes/messier/src/weights/ismp_parachain.rs [hunk 9]
```diff
@@ -0,0 +1,108 @@
+// Copyright (C) Polytope Labs Ltd.
+// SPDX-License-Identifier: Apache-2.0
```
Comment (Wizdave97) [inline parachain/runtimes/messier/src/weights/ismp_parachain.rs:2]: messier is now defunct; you can revert all changes to messier runtime.

File: modules/ismp/clients/parachain/client/src/benchmarking.rs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+	/// Parameters:
+	/// - `n`: Number of parachains to remove in a single call
+	#[benchmark]
+	fn remove_parachain(n: Linear<1, 100>) -> Result<(), BenchmarkError> {
+		let parachains: Vec<ParachainData> =
+			(0..n).map(|i| ParachainData { id: i, slot_duration: 6000u64 }).collect();
+
+		#[block]
+		{
+			Pallet::<T>::add_parachain(RawOrigin::Root.into(), parachains)?;
```
Comment (seunlanlege) [inline modules/ismp/clients/parachain/client/src/benchmarking.rs:62]: This line should be outside the block

**Review by seunlanlege — CHANGES_REQUESTED** [review]
Looks mostly good

**Review by seunlanlege — APPROVED** [review]

File: modules/ismp/core/Cargo.toml [hunk 11]
```diff
@@ -23,6 +23,7 @@ hex = { workspace = true, features = ["alloc"] }
 anyhow = { workspace = true, default-features = false }
 thiserror = { workspace = true }
 serde-hex-utils = { workspace = true, default-features = false }
+log = "0.4.22"
```
Comment (Wizdave97) [inline modules/ismp/core/Cargo.toml:26]: Please revert

Comment (Wizdave97) [inline modules/ismp/core/Cargo.toml:26]:
```suggestion
```

**Review by Wizdave97 — APPROVED** [review]

**Pull request merged**